Parameters are now associated to section in the parameter parser

julza03 opened the pull request:
The SGDIntLoadFamily_* functions, where * was a letter, were replaced… by functions where * correspond to a section name.
Thus, parameters are associated to their corresponding section.
Parameters kept for compatibility reasons are grouped in a single function

julza03 committed:
The SGDIntLoadFamily_* functions, where * was a letter, were replaced by functions where * correspond to a section name.
Thus, parameters are associated to their corresponding section.
Parameters kept for compatibility reasons are grouped in a single function

flomnes commented on src/libs/antares/study/parameters.cpp:
```diff
@@ -691,327 +791,169 @@ static bool SGDIntLoadFamily_P(Parameters& d, const String& key, const String& v
-      &SGDIntLoadFamily_Y,
-      nullptr,
-    };
-    if (c == 'h')
-    {
-        // special prototype for H
-        return SGDIntLoadFamily_H(params, key, value, rawvalue, version);
-    }
-    else
-    {
-        int i = c - 'a';
-        if (jumper[i])
-            return (*(jumper[i]))(params, key, value, version);
-    }
-    return false;
+
+
```
Rename this function to `checkKeySection(const String& s)` to check that the 1st character is in range a-z.

Call it whenever a section/property name is read.

julza03 replied:
Partially corrected here : fa68674. 
I've made the correction only for parameter names (but not for sections). Indeed, section names are already checked in the map and I'm not sure that verifying it before will make any difference in the performances of the app. 
I think that this control is useless, even for parameters, with the new implementation (which is not based on the first letter). You may want to remove this function.

flomnes replied:
I agree

flomnes commented on src/libs/antares/study/parameters.cpp:
```diff
@@ -1020,10 +962,33 @@ bool Parameters::loadFromINI(const IniFile& ini, uint version, const StudyLoadOp
+            {"advanced parameters",&SGDIntLoadFamily_AdvancedParameters},
+            {"playlist",&SGDIntLoadFamily_Playlist},
+            {"variables selection",&SGDIntLoadFamily_VariablesSelection},
+            {"seeds",&SGDIntLoadFamily_Seeds}};
```
The complete name is **seeds - mersenne twister**

julza03 replied:
OK. Commits b03bbed and b5def8c

flomnes commented on src/libs/antares/study/parameters.cpp:
```diff
@@ -1020,10 +962,33 @@ bool Parameters::loadFromINI(const IniFile& ini, uint version, const StudyLoadOp
+            {"variables selection",&SGDIntLoadFamily_VariablesSelection},
+            {"seeds",&SGDIntLoadFamily_Seeds}};
+
+    Callback* handleAllKeysInSection;
```
`Callback` is already a function pointer. This is equivalent to 
```cpp
function_t** Callback;
```
I suggest you remove one indirection level.

julza03 replied:
Ok. Commits 96a1f00 (fix 0a225bf).

flomnes commented on src/libs/antares/study/parameters.cpp:
```diff
@@ -1020,10 +962,33 @@ bool Parameters::loadFromINI(const IniFile& ini, uint version, const StudyLoadOp
     {
+        sectionName = section->name;
+        sectionName.toLower();
+        handleAllKeysInSection = &sectionAssociatedToKeysProcess[section->name]; // @TODO Must check that the function exists
```
1. Use `std::map<T>::count` to determine the existence of the key within the map.
2. Use `sectionName` for the key.
3. Use `std::map<T>::at`, it won't try to add elements to your map unlike `operator[]`. More details on [at](https://www.cplusplus.com/reference/map/map/at/) and [operator\[\]](https://www.cplusplus.com/reference/map/map/operator[]/).

julza03 replied:
Ok. Other solution adopted with try/catch block around .at use. If no section is found an exception, catched, will be raised. See the documentation if needed to confirm the try/catch block. 
See a8cb305

flomnes commented on src/libs/antares/study/parameters.cpp:
```diff
@@ -1783,7 +1756,7 @@ void Parameters::saveToINI(IniFile& ini) const
     // Seeds
     {
-        auto* section = ini.addSection("seeds - Mersenne Twister");
+        auto* section = ini.addSection("seeds - Mersenne Twister"); //@TODO : Why Capital letters are used here ?
```
I suggest you keep this as is for compatibility reasons.

Possible explanation [here](https://fr.wikipedia.org/wiki/Mersenne_Twister).

julza03 replied:
OK. See [commit ](https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/342/commits/a5a76f731abb174f417a1e70ec7f1fd996da2981)

flomnes commented on src/libs/antares/study/parameters.cpp:
```diff
@@ -1020,10 +962,33 @@ bool Parameters::loadFromINI(const IniFile& ini, uint version, const StudyLoadOp
+    String sectionName;
+    typedef bool (*Callback)(Parameters&, const String&, const String&, const String&, const String&, uint);
+
+    static std::map<String,Callback> sectionAssociatedToKeysProcess = {
```
`static const`

julza03 replied:
OK. Commit 55c18c4

flomnes reviewed (changes requested):
Thank you for your contribution !

julza03 committed:
The SGDIntLoadFamily_Seeds function was renamed to SGDIntLoadFamily_SeedsMersenneTwister

julza03 committed:
Add const to the map section <> functions. The map must not be modified by the app.

julza03 committed:
Remove TODO question. Capitals are kept for compatibility reasons. Anyway, even if here it is for the saver, capitals are not a problem for the reader since section names are converted to lowercase.

julza03 committed:
sectionName parameter is useless in SGDIntLoadFamily_* signature because of the new way to call functions (which already depends on the section name)

julza03 committed:
Fusion of SGDInyLoadFamily_Others/isDeprecated in SGDIntLoadFamily_Legacy -> only kept for compatibility reasons

julza03 committed:
Remove * on Callback* because it is already included in the typedef

julza03 committed:
replace [] by .at for finding a sectionName in the map.
Also sectionName is used insted of section->name : the first one is in lowercase, the convention adopted for this map.

julza03 committed:
Complete commit b5def8c1 : error on the section name for seeds. Corrected to seeds - mersenne twister

julza03 committed:
fix unused parameters warning : rawvalue is only used SGDIntLoadFamily_General. parameters name removed from the others.

julza03 committed:
fix possible uninitialized value (callback variable) : keyword continue added to pass to the next section when the sectionName is not recognized

julza03 committed:
Checks the first letter of the key as requested

julza03 committed:
Complete commit 0a225bfe : error while removing the * - removed h instead.

julza03 commented:
Other suggestions on this file parameters.cpp : 
- Before merging, you may want to apply a reformat code - I didn't know if my editor was conform to the rules of Antares so I didn't reformatted the code. 
- #include <stdio.h> is the C syntax. To adopt C++ style, it will be better to write : #include <cstdio> (You may check why cstdio is required, because iostream, already included, may answer to your needs. But CLion warns that these library are unused - you may want to do an analysis). 
- In SGDIntLoadFamily_Optimization, the if concerning "link-type" may be rewritten like this : 
if (key == "link-type")
    {
        CString<64, false> v = value;
        v.trim();
        v.toLower();
        if (value == "ac")
            d.linkType = ltAC;
        else
            d.linkType = ltLocal;
        return true;
    }
- As in the previous version, these comments, inLoadFromINI don't reflect the code : // We convert the key and the value into the lower case format. // Do not forget the variable `key` and `value` are identical to
            // `p->key` and `p->value` except they are already in the lower case format
Only the value is converted in lower case, I think that the key must be converted too because the comparisons in SGDIntLoadFamily are made with text in lowercase. (Maybe p->key is already in lower-case, I don't know.)
-saveToINI is a very long function. I think you may want to split it in several ones, each one in charge of saving one section.

flomnes commented on src/libs/antares/study/parameters.cpp:
```diff
@@ -1020,24 +925,57 @@ bool Parameters::loadFromINI(const IniFile& ini, uint version, const StudyLoadOp
-            if (not ParametersInternalLoadProperty(*this, p->key, value, p->value, version))
+            // Do not forget the variable `key` and `value` are identical to
+            // `p->key` and `p->value` except they are already in the lower case format
+            if (not (handleAllKeysInSection)(*this, p->key, value, p->value, version))
```
Style : parenthesis around function name are not needed.

julza03 replied:
Ok : 0b051f3

flomnes commented on src/libs/antares/study/parameters.cpp:
```diff
@@ -1020,24 +925,57 @@ bool Parameters::loadFromINI(const IniFile& ini, uint version, const StudyLoadOp
     String value;
-
+    String sectionName;
+    typedef bool (*Callback)(Parameters&, const String&, const String&, const String&, uint);
```
For clarity, I suggest you comment each parameter
Example :
```cpp
typedef bool (*Callback)(Parameters& // [out] Parameter object to load the data into
                         const String& // [in] Key, comes left to the '=' sign in the .ini file
                         const String& // [in] Lowercase value, comes right to the '=' sign in the .ini file
                                       // etc.
                        );

julza03 replied:
Ok : 0b051f3

flomnes reviewed (commented):
Formatting may be incorrect, I'll fix it before a merge.

To be added to Antares-next

julza03 committed:
Remove useless parenthesis on Callback call
Add comments on the callback type definition

flomnes commented on src/libs/antares/study/parameters.cpp:
```diff
@@ -926,7 +926,11 @@ bool Parameters::loadFromINI(const IniFile& ini, uint version, const StudyLoadOp
+            const String&, // [in] Key, comes left to the '=' sign in the .ini file
+            const String&, // [in] Lowercase value, comes right to the '=' sign in the .ini file
+            const String&, // [in] Raw value as writtent right to the '=' sign in the .ini file
+            uint); // [in] Version of the simulator (such as 710)
```
Upon closer examination, the version is read from the **study**, it is not the version of Antares Simulator.

**load.cpp**
```cpp
bool Study::internalLoadHeader(const String& path)
{
    // Header
    buffer.clear() << path << SEP << "study.antares";
    if (not header.loadFromFile(buffer))
    {
        logs.error() << path << ": impossible to open the header file";
        return false;
    }
// ...
}
```
**header.cpp**
```cpp
bool StudyHeader::internalLoadFromINIFile(const IniFile& ini, bool warnings)
{
    version = 0;
    const IniFile::Section* sect = ini.find("antares");
// ...
            // Version
            if (p->key == "version")
            {
                version = p->value.to<uint>(); // Property "version" in section "antares" from file study.antares
                continue;
            }
// ...
}
```

flomnes reviewed (commented):
One last change: comment at line 933

flomnes committed:
Apply clang-format and change one comment

antares version -> study version l. 943
